### PR TITLE
GradModeVariable should guard on the initial grad state

### PR DESCRIPTION
**Pull request by yanboliang**
This bug was found when I run training loss benchmark, where ```GRAD_MODE``` guard keeps failing and hits cache limit.

```GradModeVariable``` has a singleton guard on the grad state, but the state was decided when ```GuardBuilder.GRAD_MODE``` was called. It was after the code object transformation, where the grad state may be changed. 

The fix is to use the grad state at the entry of the frame execution.

### CODE REVIEW

File: torchdynamo/variables/misc.py [hunk 1]
```diff
@@ -262,6 +257,10 @@ def call_function(
 
 
 class GradModeVariable(ContextWrappingVariable):
+    """represents torch.{no_grad,enable_grad,set_grad_mode}()"""
+
+    _guards_singleton = {Guard("", GuardSource.GLOBAL, GuardBuilder.GRAD_MODE)}
+
```
Comment (yanboliang) [inline torchdynamo/variables/misc.py:263]: code refactor, they are relevant to grad mode only.

File: torchdynamo/eval_frame.py [hunk 2]
```diff
@@ -190,6 +191,8 @@ class OptimizeContext(_TorchDynamoContext):
     def __init__(self, callback, backend_ctx_ctor):
         def on_enter():
             global most_recent_backend
+            global initial_grad_state
+            initial_grad_state = torch.is_grad_enabled()
```
[the remark this answers is not in the document]
Comment (yanboliang) [reply torchdynamo/eval_frame.py:195]: Oh, yes, you are right!

**Review by anijain2305 — APPROVED** [review]
LGTM. But, lets wait for @jansel review to be sure.

File: torchdynamo/convert_frame.py [hunk 3]
```diff
@@ -72,6 +72,9 @@ def clear(self):
 output_codes = Tracker()
 
 
+initial_grad_state = None
```
Comment (jansel) [inline torchdynamo/convert_frame.py:75]: Ideally we wouldn't use a global for this.

**Review by jansel — APPROVED** [review]

File: torchdynamo/convert_frame.py (hunk 3 above)
Comment (yanboliang) [reply torchdynamo/convert_frame.py:75]: I'll address this as follow-up, thanks!

File: torchdynamo/convert_frame.py [hunk 4]
```diff
@@ -316,6 +319,9 @@ def format_guard_failures(code):
         if not has_tensor_in_frame(frame):
             return None
 
+        global initial_grad_state
+        initial_grad_state = torch.is_grad_enabled()
```
Comment (voznesenskym) [inline torchdynamo/convert_frame.py:323]: It's odd to me that we special case this, and to @jansel's point, not great that it's global - what we should have instead is a state snapshot that is passed around. Grad is just one such state.

File: torchdynamo/guards.py [hunk 5]
```diff
@@ -404,11 +406,11 @@ def OBJECT_MUTATION(self, guard: Guard):
         mutation_guard.watch(self.get(guard.name), self.guarded_code)
 
     def GRAD_MODE(self, guard: Guard):
-        """Guard on the current value of torch.is_grad_enabled()"""
+        """Guard on the initial grad state"""
         assert guard.name == ""
         assert guard.source is GuardSource.GLOBAL
         code = None
-        if torch.is_grad_enabled():
```
Comment (voznesenskym) [inline torchdynamo/guards.py:411]: Can you explain this to me a little more? When would torch.is_grad_enabled() not line up with the state stored in `convert_frame.initial_grad_state` - is it because someone swapped grad state, but it happened in a skip / graph break? If so, shouldn't we actually be checking the current torch.is_grad_enabled()? Why have they deviated?

Comment (yanboliang) [reply torchdynamo/guards.py:411]: Take a look at the unit test I added, if grad_enabled is True before the ```torch.set_grad_enabled``` and user's code sets it to ```False``` before graph break, then when this guard creating, the state is False which is not line up with the state stored in ```convert_frame.initial_grad_state``` which is True.

File: torchdynamo/convert_frame.py (hunk 4 above)
Comment (yanboliang) [reply torchdynamo/convert_frame.py:323]: I'll address the global variable issue in a follow up PR. We are planning to publish the E2E training performance result, which is blocked by this PR.

File: torchdynamo/guards.py (hunk 5 above)
Comment (voznesenskym) [reply torchdynamo/guards.py:411]: But if user code says grad is false, why are we using true?

Comment (yanboliang) [reply torchdynamo/guards.py:411]: We are guarding on the enter state, not the exit state, right?

Comment (voznesenskym) [reply torchdynamo/guards.py:411]: Yes, enter state. 

If the user code says grad is set to false, we should reflect that in the guard.

What I don't understand is how torch.is_grad_enabled() can give us the wrong value for the state of grad... Why do we want the initial value, instead of the user set one?
